src/types.ts

```diff
@@ -152,24 +157,31 @@ export interface BinaryRecordedEvent extends RecordedEventBase {
   metadata: Uint8Array;
 }
 
-export interface AllStreamJSONRecordedEvent extends JSONRecordedEvent {
+export interface AllStreamJSONRecordedEvent<
+  EventType extends string = string,
+  Data = unknown,
+  Metadata extends Record<string, unknown> = Record<string, unknown>
+> extends JSONRecordedEvent<EventType, Data, Metadata> {
   /**
    * Position of this event in the transaction log.
    */
   position: Position;
 }
 
-export interface AllStreamBinaryRecordedEvent extends BinaryRecordedEvent {
+export interface AllStreamBinaryRecordedEvent<EventType extends string = string>
+  extends BinaryRecordedEvent<EventType> {
   /**
    * Position of this event in the transaction log.
    */
   position: Position;
 }
 
-export type RecordedEvent = JSONRecordedEvent | BinaryRecordedEvent;
-export type AllStreamRecordedEvent =
-  | AllStreamJSONRecordedEvent
-  | AllStreamBinaryRecordedEvent;
+export type RecordedEvent<EventType extends string = string> =
+  | JSONRecordedEvent<EventType>
+  | BinaryRecordedEvent<EventType>;
+export type AllStreamRecordedEvent<EventType extends string = string> =
+  | AllStreamJSONRecordedEvent<EventType>
+  | AllStreamBinaryRecordedEvent<EventType>;
```

Comment:
This is very nit-picky, but I'm not sure if there is value in making `RecordedEvent` and `AllStreamRecordedEvent` generic in this way, and it may add confusion having it as an option.

My reasoning is
``` ts
RecordedEvent<'my-event' | 'my-other-event'>
```
gives no more valuable information than
``` ts
RecordedEvent
```
as we still dont know if it is a json event, and what we expect its payload to be etc, so discriminating the events on `.eventType` doesnt help us. 

This type is more of a "could be any event that came from a stream / $all stream.

(More than happy to change my mind if there is something I'm missing here)

Reply:
Just to expand on that:

If someone were to do:


```ts
type EventX = RecordedEvent<'eventX'>;
type EventY = RecordedEvent<'eventY'>;
type EventZ = RecordedEvent<'eventZ'>;

connection.subscribeToStream<EventX | EventY | EventZ>()
```

it is _probably_ not what they were intending to do, and if it was, then they could do:

```ts
type EventX = JSONRecordedEvent<'eventX', { some: string }, { meta: string }>;
// eventY could be absolutely anything:
type EventY = JSONRecordedEvent<'eventY'> | BinaryRecordedEvent<'eventX'>; 
type EventZ = JSONRecordedEvent<'eventZ', { z: string }, { meta: string }>;

connection.subscribeToStream<EventX | EventY | EventZ>()
```

so I dont think it should be an option to do, to stop people making the mistake.

Reply:
Yeah, I can agree with that for the `$all` stream. For arbitrary streams, I'm looking for a way to express "it's either a binary event with `eventType` `x`, `y` or `z`, or a JSON event with `eventType` `l`, `m` or `n`". But I'm not seeing a good non-confusing way to do that at the moment, so I'll leave it. Maybe a follow-up PR.

Reply:
I fixed this in 04e119eb571f709e68aaa98bc7d2e9c3b08f9161